### ISSUE TITLE
DLL Load Failed and possible solution?

### ISSUE THREAD

**Issue by devaku** [PyGreSQL/PyGreSQL#53]
Hey guys, I'm posting this here in hopes that this gets fixed as soon as possible because this is a glaring problem.

I followed the tutorial here:
https://pygresql.org/contents/tutorial.html

```
from pg import DB

import os
from dotenv import load_dotenv

# Load the ENV
load_dotenv()

dbname = os.getenv('DB_NAME')
host = os.getenv('DB_HOST')
port = os.getenv('DB_PORT')
user = os.getenv('DB_USERNAME')
password = os.getenv('DB_PASSWORD')

# Initialize the database
db = DB(dbname, host, port, user, password)
```

And when I tried to run it, I ran into the dreaded: **DLL load failed: The operating system cannot run %1**

After troubleshooting it myself, I've noticed that the code only grabs the file paths in your PATHs in your system settings, and if the folder has "libpq.dll". It has no regard whatsoever if it's in the right folder to begin with--if it's inside a PostgresSQL folder.

Unfortunately, in my PATHs setting, I had _C:\xampp\php_ BEFORE _C:\Program Files\PostgreSQL\12\bin\_ which caused quite a problem, as PHP has its own version of "libpq.dll".

Long story short, the problem was easily fixed by moving the Postgres path before the xampp path and the application threw no error.

Is it possible to have an extra "if" in the source code to address this matter?

**Comment by Cito** [comment]
Hi @devaku, thanks for reporting. You say "It has no regard whatsoever if it's in the right folder".

The problem here is how you define *right* folder. Maybe in another situation, the dll from xampp *is* the current dll you want to use, and the one in the PostgreSQL directory is from a left-over installation of an old PostgreSQL version?

So far, we consider what is *right* by the ordering of the PATH. If we change that order, then we might go against the explicit wish of the user who might have a good reason for ordering the PATH that way. Furthermore, the path name where Postgres is installed is not standardized, it must not be "C:\Program Files\PostgreSQL\12" even on Windows. And you could also have two PostgreSQL versions installed alongside, maybe some inactive ones etc. From looking at the path alone, it is therefore impossible to know which one is "right".

But you're still giving me an idea here. Maybe we can make better use of the new `os.add_dll_directory` function that is new in Python 3.8. A possible mitigation could be to try importing the DLLs in the order from the PATH by using said function, ignoring old or bad DLLs that cannot be loaded, and if no loadable DLL can be found, print a better error message explaining the user what to do. That could work for Python >= 3.8. 

Will try to experiment with this idea later this week.

**Comment by devaku** [comment]
You wrote it better than I could put it in words.

That said, I'm surprise that it requires an outside DLL like it to work. I happen to have Postgres installed on my PC because I need it. Shouldn't it come with its own DLL in the event that the user doesn't have one installed on their own PC? 

I didn't expect it to have this requirement because the instructions on the app just told me to install it via pip. I thought it just worked right outside the box.

**Comment by Cito** [comment]
Maybe we should make it more clear in the documentation then.

You don't need to have a full PostgreSQL installation on your machine. You just need the libpq client library. On Unix this usually comes as a separately installable package.

On Windows, we could consider bundling an appropriate dll with the binaries (added as issue #54 now). But that's not a simple task even if the license are compatible and allow this, and I currently do not have enough time and energy to work on this.

**Comment by Cito** [comment]
@devaku Your use case should now be covered with the latest change. Will be released in 5.2.2.
